subsys/bluetooth/host/hci_core.c

```diff
@@ -853,41 +853,61 @@ static void enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
 	}
 #endif
 
-	if (IS_ENABLED(CONFIG_BT_CENTRAL)) {
-		if (evt->status) {
-			/*
-			 * if there was an error we are only interested in pending
-			 * connection so there is no need to check ID address as
-			 * only one connection can be in that state
-			 *
-			 * Depending on error code address might not be valid anyway.
-			 */
-			conn = find_pending_connect(NULL);
-			if (!conn) {
-				return;
-			}
-
-			conn->err = evt->status;
-
+	if (evt->status) {
 		/*
-		 * Handle advertising timeout after high duty directed
-		 * advertising.
+		 * If there was an error we are only interested in pending
+		 * connection. There is no need to check ID address as
+		 * only one connection can be in that state.
+		 *
+		 * Depending on error code address might not be valid anyway.
 		 */
-		if (conn->err == BT_HCI_ERR_ADV_TIMEOUT) {
-			atomic_clear_bit(bt_dev.flags, BT_DEV_ADVERTISING);
+		conn = find_pending_connect(NULL);
+		if (!conn) {
+			return;
 		}
 
-			bt_conn_set_state(conn, BT_CONN_DISCONNECTED);
+		conn->err = evt->status;
 
-			/* check if device is market for auto connect */
-			if (atomic_test_bit(conn->flags, BT_CONN_AUTO_CONNECT)) {
-				bt_conn_set_state(conn, BT_CONN_CONNECT_SCAN);
+		if (IS_ENABLED(CONFIG_BT_PERIPHERAL)) {
+			/*
+			 * Handle advertising timeout after high duty directed
+			 * advertising.
+			 */
+			if (conn->err == BT_HCI_ERR_ADV_TIMEOUT) {
+				atomic_clear_bit(bt_dev.flags,
+						 BT_DEV_ADVERTISING);
+				bt_conn_set_state(conn, BT_CONN_DISCONNECTED);
+
+				goto done;
 			}
+		}
 
-			goto done;
+		if (IS_ENABLED(CONFIG_BT_CENTRAL)) {
+			/*
+			 * Handle cancellation of outgoing connection attempt.
+			 */
+			if (conn->err == BT_HCI_ERR_UNKNOWN_CONN_ID) {
+				/* We notify before checking autoconnect flag
+				 * as application may choose to change it from
+				 * callback.
+				 */
+				bt_conn_set_state(conn, BT_CONN_DISCONNECTED);
+
+				/* Check if device is marked for autoconnect. */
+				if (atomic_test_bit(conn->flags,
+						    BT_CONN_AUTO_CONNECT)) {
+					bt_conn_set_state(conn,
+							  BT_CONN_CONNECT_SCAN);
+				}
+
+				goto done;
+			}
 		}
-	} else {
-		BT_ASSERT(evt->status == 0);
+
+		/* Other errors unallowed. */
+		BT_ASSERT(false);
```

Comment:
This has me a little bit nervous. Are we really sure that there's no other status codes that we can get, i.e. this has been double-checked with the HCI spec.

Also, we generally reserve asserts only for things which indicate a bug in *our* code. I.e. buggy behavior in an external component (the controller in this case) would only result in an error or ignoring the data on our side.

Unrelated to the above (and this PR) it'd be nice to have a BT_ASSERT_NOT_REACHED() that'd do the same as BT_ASSERT(false) but be much clearer in its intent for the casual reader.

Reply:
@sjanc , any idea?

Reply:
@pdunaj sorry to keep dragging this PR on - I also want to see it merged ASAP, however this is really something that'd be good to sort out first. What I'd suggest is to get rid of the assert and instead have the following kind of branches in two places for the specific errors, but let the code otherwise continue executing even if we get an unexpected error:
```
if (evt->status != SOME_EXPECTED_ERROR) {
        BT_WARN("Got unexpected status 0x%02x", evt->status);
}
```

Reply:
I don't think w can get other error codes but yeah, buggy controller could trigger it. So for other error code we could warn and just drop conn. I initially added assert since in case of failure error code indicate which connection (slave or master) failed. So for other error we don't know whats the expected outcome (eg was advertising stopped? etc)

Reply:
Warning added instead of assert.